test/v2/controllers/CreateAmendForeignPropertyAnnualSubmissionControllerSpec.scala

```diff
@@ -125,108 +66,90 @@ class CreateAmendForeignPropertyAnnualSubmissionControllerSpec
-          (InternalError, INTERNAL_SERVER_ERROR),
-          (RuleIncorrectGovTestScenarioError, BAD_REQUEST)
-        )
+    "service errors occur" should {
+      "the service returns an error" in new Test {
+        MockAmendForeignPropertyAnnualSubmissionRequestParser
+          .parseRequest(rawData)
+          .returns(Right(request))
+
+        MockAmendForeignPropertyAnnualSubmissionService
+          .amend(request)
+          .returns(Future.successful(Left(ErrorWrapper(correlationId, InternalError))))
 
-        input.foreach(args => (serviceErrors _).tupled(args))
+        runErrorTest(InternalError)
       }
     }
   }
 
+  trait Test extends ControllerTest with AuditEventChecking[GenericAuditDetail] {
+
+    val controller = new CreateAmendForeignPropertyAnnualSubmissionController(
+      authService = mockEnrolmentsAuthService,
+      lookupService = mockMtdIdLookupService,
+      parser = mockCreateAmendForeignPropertyAnnualSubmissionRequestParser,
+      service = mockService,
+      auditService = mockAuditService,
+      hateoasFactory = mockHateoasFactory,
+      cc = cc,
+      idGenerator = mockIdGenerator
+    )
+
+    protected def callController(): Future[Result] = controller.handleRequest(nino, businessId, taxYear)(fakeRequestWithBody(requestJson))
+
+    protected val requestJson: JsValue = createAmendForeignPropertyAnnualSubmissionRequestBodyMtdJson
+
+    protected val testHateoasLink: Link = Link(href = s"/individuals/business/property/$nino/$businessId/annual/$taxYear", method = GET, rel = "self")
+
+    protected val hateoasResponse: JsValue = Json.parse(
+      s"""
+         |{
+         |   "links": [
+         |      {
+         |         "href": "/individuals/business/property/$nino/$businessId/annual/$taxYear",
+         |         "method": "GET",
+         |         "rel": "self"
+         |      }
+         |   ]
+         |}
+    """.stripMargin
+    )
+
+    protected def event(auditResponse: AuditResponse, maybeRequestBody: Option[JsValue]): AuditEvent[GenericAuditDetail] =
+      AuditEvent(
+        auditType = "CreateAmendForeignPropertyAnnualSubmission",
+        transactionName = "create-amend-foreign-property-annual-submission",
+        detail = GenericAuditDetail(
+          versionNumber = "2.0",
+          userType = "Individual",
+          agentReferenceNumber = None,
+          params = Json.obj("nino" -> nino, "businessId" -> businessId, "taxYear" -> "2019-20", "request" -> requestJson),
+          correlationId = correlationId,
+          response = auditResponse
+        )
+      )
+
+    val body: CreateAmendForeignPropertyAnnualSubmissionRequestBody = createAmendForeignPropertyAnnualSubmissionRequestBody
+
+    protected val rawData: CreateAmendForeignPropertyAnnualSubmissionRawData =
+      CreateAmendForeignPropertyAnnualSubmissionRawData(nino, businessId, taxYear, requestJson)
+
+    protected val request: CreateAmendForeignPropertyAnnualSubmissionRequest =
```

Comment:
would be good to rename `request` to `requestData` so that we have all our data vals together making it a bit more readable (so we would have rawData, requestData, hateoasData). Could also be worth renaming `body` to `requestBody` to make it clear that its for the request and not response. This comment would apply to the other controller specs